File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/RoleCreateRequestProto.lua
--创建角色(请求)
RoleCreateRequestProto = { ProtoCode = 10202, JobId = 0, RoleName = "" }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleCreateRequestProto.__index = RoleCreateRequestProto;

function RoleCreateRequestProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleCreateRequestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleCreateRequestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteByte(proto.JobId);
    ms:WriteUTF8String(proto.RoleName);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleCreateRequestProto.GetProto(buffer)

    local proto = RoleCreateRequestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.JobId = ms:ReadByte();
    proto.RoleName = ms:ReadUTF8String();

    ms:Dispose();
    return proto;
end

File: MMORPG_2/Assets/Script/Data/Proto/CreateProto/RoleCreateRequestProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:21
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 创建角色(请求) </summary>
public struct RoleCreateRequestProto : IProto
{
    public ushort ProtoCode { get { return 10202; } }

    /// <summary> 职业Id </summary>
    public byte JobId;
    /// <summary> 角色名称 </summary>
    public string RoleName;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteByte(JobId);
            ms.WriteUTF8String(RoleName);
            return ms.ToArray();
        }
    }

    public static RoleCreateRequestProto GetProto(byte[] buffer)
    {
        RoleCreateRequestProto proto = new RoleCreateRequestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.JobId = (byte)ms.ReadByte();
            proto.RoleName = ms.ReadUTF8String();
        }
        return proto;
    }
}